### ISSUE TITLE
[Security] Workflow `docs-preview-deploy.yml` is using vulnerable action `nwtgck/actions-netlify`

### ISSUE THREAD

**Issue by qazaqstan2025** [docker-mailserver/docker-mailserver#2332]
The workflow docs-preview-deploy.yml is referencing action nwtgck/actions-netlify using references v1.2. However this reference is missing the commit [223b5b5981680adfe8ec4c9f471620cbbdbfadec](https://github.com/nwtgck/actions-netlify/commits/223b5b5981680adfe8ec4c9f471620cbbdbfadec) which may contain fix to the some vulnerability. 
The vulnerability fix that is missing by actions version could be related to: 
(1) CVE fix 
(2) upgrade of vulnerable dependency 
(3) fix to secret leak and others. 
Please consider to update the reference to the action.

**Comment by georglauterbach** [comment]
Thank you for reporting this vulnerability. We will look into it. I have assigned @polarathene too as he created the documentation deployment. @polarathene can you have a look here?

Moreover, with the linting-update-PR #2329, I run the GitHub Super Linter over this repo to check the GitHub actions too. There was one more report:

For the line

https://github.com/docker-mailserver/docker-mailserver/blob/99cc9fec2a01e078fda6ce439def681e31a4e8f9/.github/workflows/docs-preview-prepare.yml#L61

the `${{ github.event.pull_request.title }}` was reported as "untrusted" and we ought to use a dedicated variable... I'm not 100% sure but I guess this is because the creator of the PR can choose the title and perform a code injection? This issue might be related: https://github.com/rhysd/actionlint/issues/19. We should do something about this as well.

**Comment by polarathene** [comment]
> I have assigned @polarathene too as he created the documentation deployment. @polarathene can you have a look here?

The action has not published a newer release officially, `1.2` would use the latest `1.2.2` release from May, prior to the referenced commit. It was not immediately clear to me from the report what the vulnerability being reported to us was however.

If it was [the one patched here](https://github.com/nwtgck/actions-netlify/pull/572), this is nothing for us to be concerned about.

> Vulnerabilities fixed
> ReDoS in Sec-Websocket-Protocol header
> Impact
> A specially crafted value of the Sec-Websocket-Protocol header can be used to significantly slow down a ws server.

We're just deploying to Netlify from GH Actions, no websocket server performance concerns.

---

> I'm not 100% sure but I guess this is because the creator of the PR can choose the title and perform a code injection?

That line is from the prepare workflow which runs in an untrusted context anyway, a security feature / requirement by Github for PR triggered workflows (at least those by third-party untrusted contributors). I don't see how that would cause any concerns, but I'm not a shell expert.

We capture the title from Github and store it as an ENV in an env file to read in the 2nd part of the workflow (deployment, which is in a trusted context with access to our secrets), but at this point that execution should have already happened with echo? We then read it during deployment to form the deployment message.

---

Unless someone knows of either having a practical vulnerability to exploit, please share, but AFAIK, we're ok and neither seems to be a vector of concern?

**Comment by georglauterbach** [comment]
> > I'm not 100% sure but I guess this is because the creator of the PR can choose the title and perform a code injection?
> 
> That line is from the prepare workflow which runs in an untrusted context anyway, a security feature / requirement by Github for PR triggered workflows (at least those by third-party untrusted contributors). I don't see how that would cause any concerns, but I'm not a shell expert.
>
> We capture the title from Github and store it as an ENV in an env file to read in the 2nd part of the workflow (deployment, which is in a trusted context with access to our secrets), but at this point that execution should have already happened with echo? We then read it during deployment to form the deployment message.

Seems valid to me. So using `" ; cat .git/config|base64 ; echo "` should not affect us, right?

**Comment by polarathene** [comment]
> Seems valid to me. So using `" ; cat .git/config|base64 ; echo "` should not affect us, right?

I don't think so, ~~but best way to confirm is probably to create a throwaway repo that tests such. Or ask on the GH Actions community forum/support~~ (**EDIT:** reviewed the flow of the PR title and official workflow/actions docs, pretty sure we're fine).

https://github.com/docker-mailserver/docker-mailserver/blob/99cc9fec2a01e078fda6ce439def681e31a4e8f9/.github/workflows/docs-preview-prepare.yml#L61

Github pre-processes job `input`s (parameters/args) before handing them over to the GH Action being called, so any content within `${{ ... }}` is interpolated as an expression `${{ ... }}` (_[documented here](https://docs.github.com/en/actions/learn-github-actions/expressions)_). We then only use it like so:

https://github.com/docker-mailserver/docker-mailserver/blob/99cc9fec2a01e078fda6ce439def681e31a4e8f9/.github/workflows/docs-preview-deploy.yml#L75

And this is an `input` defined under [`job.with`](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idwith), which the action internally accesses via [this JS API](https://github.com/nwtgck/actions-netlify/blob/bda6b2e6da6fdf666e0accd8873922698c37974d/src/inputs.ts#L29-L31):

```JS
deployMessage() {
  return core.getInput('deploy-message') || undefined
},
```

At which point the input string was interpolated already.

---

As the previous docs link notes, the PR title could be exploited to execute code, and the official GH Actions/workflow docs [educate us on that topic here](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections) with some examples.

So it'd seem that could be taken advantage of in the prepare workflow where we do run the shell script, but there shouldn't be any access to secrets in that workflow, and the deploy workflow will just print the PR title as part of the input string, not able to execute anything within the secure context.

Still, we could better approach this I guess by using the expressions to set temporary ENV for the shell command that we then write to a file? (eg: `echo "PR_TITLE=${PR_TITLE}"` is presumably safer, and `echo "PR_TITLE=${{env.PR_TITLE}}"` would be just as bad as the current interpolation expression since it injects/replaces with it's content rather than behaving like shell vars?).

- https://docs.github.com/en/actions/learn-github-actions/environment-variables
- https://docs.github.com/en/actions/learn-github-actions/workflow-commands-for-github-actions#setting-an-environment-variable

Documentation seems to advise using the ENV var directly when running shell commands, and only the `env` context for expressions elsewhere.

---

@qazaqstan2025 can confirm if I'm mistaken about the vulnerability with `nwtck/actions-netlify` being about the the WebSocket server perf exploit, which doesn't affect us.

It does look like that action ~~isn't being actively maintained anymore (swamped with PRs from dependabot, some merged months ago but no new releases made)~~ nevermind, the maintainer just released 1.2.3 today with the referenced commits, so regardless we're good once we update!

## TL;DR

I think we're good, but just in case, the prepare workflow can better handle collecting the PR title to store as an ENV. At present the only exploit would be to run shell commands in the untrusted context where it shouldn't be able to accomplish anything, no concerns with it's usage in the secure context for the deploy workflow.

`nwtgck/actions-netlify` just released 1.2.3, so that can be resolved too.

**Comment by polarathene** [comment]
This issue has been resolved via PR #2337 :+1: 

Thanks for the report @qazaqstan2025 , much appreciated!
